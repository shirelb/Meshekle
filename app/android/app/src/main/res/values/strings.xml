<resources>
    <string name="app_name">משקלה</string>
    <string name="default_notification_channel_id" translatable="false">fcm_default_channel</string>
</resources>
