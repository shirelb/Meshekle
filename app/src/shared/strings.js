const loginScreenStrings = {
    LOGIN: "התחבר",
    WELCOME_TO_LOGIN: "Welcome to the login screen!",
    USER_ID_PLACEHOLDER: "ת.ז.",
    PASSWORD_PLACEHOLDER: "סיסמא",
    APP_NAME: "משקל\'ה",
    WRONG_CREDENTIALS: "אחד הפרטים שהזנת לא נכון",
    NO_SUCH_USER: "המשתמש אינו קיים",
    EMPTY_USER_ID: "הוסף ת.ז. ונסה להתחבר שוב",
    EMPTY_PASSWORD: "הוסף סיסמא ונסה להתחבר שוב",
};

const mainScreenStrings = {
    LOGOUT: "התנתק",
    WELCOME: "ברוכים הבאים ",
};

const drawerMenu = {
    LOGOUT: "התנתק",
    MAIN_SCREEN_NAME: "בית",
    APPOINTMENTS_SCREEN_NAME: "תורים",
    CHORES_SCREEN_NAME: "תורנויות",
    PhoneBook_SCREEN_NAME: "ספר טלפונים",
};

export default {loginScreenStrings,mainScreenStrings,drawerMenu};