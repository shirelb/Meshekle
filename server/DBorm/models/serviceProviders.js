module.exports = (sequelize, type) => {
    return sequelize.define('ServiceProviders', {
        serviceProviderId:{
            type: type.INTEGER,
            primaryKey: true,
            autoIncrement: true,
            allowNull: false,
            required: true
        },
        role : {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        userId: {
            type: type.INTEGER,
            allowNull: false,
            required: true
        },
        operationTime: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        phoneNumber: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
    });
};