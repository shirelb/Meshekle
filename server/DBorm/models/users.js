module.exports = (sequelize, type) => {
    return sequelize.define('Users', {
        userId:{
            type: type.STRING,
            primaryKey: true,
            required: true
        },
        fullname: {
            type: type.STRING,
            required: true
        },
        password: {
            type: type.STRING,
            required: true
        },
        email: {
            type: type.STRING,
            required: true
        },
        mailbox: {
            type: type.INTEGER,
            required: true
        },
        cellphone: {
            type: type.STRING,
            required: true
        },
        phone: {
            type: type.STRING
        },
        bornDate: {
            type: type.DATE,
            required: true
        },
        active: {
            type: type.BOOLEAN,
            required: true,
            defaultValue: true
        },
        image: {
            type: type.BLOB,
        }
    });
};