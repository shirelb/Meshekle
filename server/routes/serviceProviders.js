var express = require('express');
var router = express.Router();
const Sequelize = require('sequelize');
const Op = Sequelize.Op;
const {Users, ServiceProviders, AppointmentRequests, Appointments, Incidents, Events} = require('../DBorm/DBorm');

/* GET users listing. */
router.get('/', function(req, res, next) {
  res.send('respond with a resource serviceProviders');
});


/* POST user . */
router.post('/users/add', function (req, res, next) {
    Users.create({
        userId: req.body.userId,
        fullname: req.body.fullname,
        password: req.body.password,
        email: req.body.email,
        mailbox: req.body.mailbox,
        cellphone: req.body.cellphone,
        phone: req.body.phone,
        bornDate: new Date(req.body.bornDate),
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

/* POST user . */
router.post('/add', function (req, res, next) {
    ServiceProviders.create({
        serviceProviderId: req.body.serviceProviderId,
        userId: req.body.userId,
        role: req.body.role,
        operationTime: req.body.operationTime,
        phoneNumber: req.body.phoneNumber,
        appointmentWayType: req.body.appointmentWayType,
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

module.exports = router;
