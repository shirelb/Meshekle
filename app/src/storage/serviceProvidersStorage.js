import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getServiceProviders = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/serviceProviders`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('load appointments error ', error)
        });
};


export default {getServiceProviders};
