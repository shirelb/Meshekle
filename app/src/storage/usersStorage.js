import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getUserEvents = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/events/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user events error ', error)
        });
};

var getUserById = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var getUsers = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/users`,
        {headers: userHeaders}
    )
        .then(response => {
            return response.data.filter(user => user.userId !== 1 && user.userId !== "1");
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var userValidToken = function (token) {
    return axios.post(`${SERVER_URL}/api/users/validToken`,
        {
            "token": token,
        },
    );
}

var userLogin = function (userId, password) {
    return axios.post(`${SERVER_URL}/api/users/login/authenticate`,
        {
            "userId": userId,
            "password": password
        },
    );
}


export default {getUserEvents, getUserById, userValidToken, userLogin, getUsers};
