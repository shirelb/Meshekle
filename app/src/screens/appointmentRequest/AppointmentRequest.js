import React, {Component} from 'react';
import {FlatList, StyleSheet, View} from 'react-native';
import {Icon, List, ListItem, SearchBar} from 'react-native-elements';
import phoneStorage from "react-native-simple-store";
import AppointmentRequestForm from "../../components/appointmentRequestForm/AppointmentRequestForm";
import serviceProvidersStorage from "../../storage/serviceProvidersStorage";
import usersStorage from "../../storage/usersStorage";


export default class AppointmentRequest extends Component {
    constructor(props) {
        super(props);

        this.state = {
            serviceProviders: [],
            formModal: false,
            serviceProviderSelected: {},
            noServiceProviderFound: false,
        };

        this.serviceProviders = [];
        this.requestAppointment = this.requestAppointment.bind(this);
    }

    componentDidMount() {
        phoneStorage.get('userData')
            .then(userData => {
                console.log('agenda componentDidMount userData ', userData);
                this.userHeaders = {
                    'Authorization': 'Bearer ' + userData.token
                };
                this.userId = userData.userId;
                this.loadServiceProviders();
            });
    }

    loadServiceProviders() {
        serviceProvidersStorage.getServiceProviders(this.userHeaders)
            .then((response) => {
                let serviceProviders = response.data;

                serviceProviders.forEach(provider => {
                    usersStorage.getUserById(provider.userId, this.userHeaders)
                        .then(user => {
                            provider.fullname = user.data[0].fullname;

                            this.setState({
                                serviceProviders: serviceProviders,
                            });

                            this.serviceProviders = serviceProviders;
                        })
                        .catch(error => {
                            console.log('error ', error)
                        });
                })
            });
    };

    requestAppointment = (serviceProvider) => {
        this.setState({
            formModal: true,
            serviceProviderSelected: serviceProvider
        });
        console.log('pressed on serviceProvider ', this.state.formModal, this.state.serviceProviderSelected);
    };

    renderSeparator = () => {
        return (
            <View
                style={{
                    height: 1,
                    width: "86%",
                    backgroundColor: "#CED0CE",
                    marginLeft: "14%"
                }}
            />
        );
    };

    updateSearch = search => {
        console.log("in search ", search);
        // this.setState({search});
        let searchText = search.toLowerCase();
        let serviceProviders = this.state.serviceProviders;
        let filteredByNameOrRole = serviceProviders.filter((item) => {
            return item.fullname.toLowerCase().match(searchText) || item.role.toLowerCase().match(searchText);
        });
        // let filteredByRole = serviceProviders.filter((item) => {
        //     return item.role.toLowerCase().match(searchText)
        // });
        if (!searchText || searchText === '') {
            this.setState({
                serviceProviders: this.serviceProviders
            })
        } else if (!Array.isArray(filteredByNameOrRole) && !filteredByNameOrRole.length) {
            // set no data flag to true so as to render flatlist conditionally
            this.setState({
                noServiceProviderFound: true
            })
        } else if (Array.isArray(filteredByNameOrRole)) {
            this.setState({
                noServiceProviderFound: false,
                serviceProviders: filteredByNameOrRole
            })
        }
    };

    renderHeader = () => {
        return <SearchBar
            placeholder="רשום פה..."
            lightTheme
            onChangeText={this.updateSearch.bind(this)}
            // round
        />;
    };

    renderRow = ({item}) => {
        return (
            <ListItem
                roundAvatar
                title={item.role}
                subtitle={item.fullname}
                // avatar={{uri:item.avatar_url}}
                onPress={() => this.requestAppointment(item)}
                containerStyle={{borderBottomWidth: 0}}
                rightIcon={<Icon name={'chevron-left'}/>}
            />
        )
    };

    render() {
        return (
            <View>
                <List containerStyle={{borderTopWidth: 0, borderBottomWidth: 0}}>
                    {this.state.noServiceProviderFound ? <Text>לא נמצאו תוצאות</Text> :
                        <FlatList
                            data={this.state.serviceProviders}
                            renderItem={this.renderRow}
                            keyExtractor={item => item.userId}
                            ItemSeparatorComponent={this.renderSeparator}
                            ListHeaderComponent={this.renderHeader}
                        />
                    }
                </List>
                <AppointmentRequestForm
                    modalVisible={this.state.formModal}
                    serviceProvider={this.state.serviceProviderSelected}
                />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    }
});

