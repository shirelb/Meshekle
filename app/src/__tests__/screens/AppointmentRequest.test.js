import React from 'react';

import {FlatList} from 'react-native';
import {List} from 'react-native-elements';

import {shallow} from "enzyme/build";
import serviceProvidersStorage from "../../storage/serviceProvidersStorage";
import usersStorage from "../../storage/usersStorage";
import users from "../jsons/users";
import serviceProviders from "../jsons/serviceProviders";
import AppointmentRequest from "../../screens/appointmentRequest/AppointmentRequest";
import phoneStorage from "react-native-simple-store";
import strings from "../../shared/strings";

jest.mock("react-native-simple-store");
jest.mock("../../storage/usersStorage");
jest.mock("../../storage/serviceProvidersStorage");


describe("AppointmentRequest should", () => {
    let wrapper = null;
    let componentInstance = null;
    const props = {};
    const serviceProviderTest = serviceProviders[41];
    const mockStore = {
        userData: {
            userId: "549963652",
            token: "some token"
        }
    };
    const navigation = {
        navigate: jest.fn(),
        state: {
            params: {
                selectedDate: "2019-06-03"
            }
        }
    };

    phoneStorage.get = jest.fn().mockImplementation((key) => Promise.resolve(mockStore[key]));
    usersStorage.getUsers = jest.fn().mockImplementation(() => Promise.resolve(users));
    usersStorage.getUserByUserID = jest.fn().mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviders = jest.fn().mockImplementation(() => Promise.resolve(serviceProviders));
    serviceProvidersStorage.getServiceProviderUserDetails = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));

    let appointmentsServiceProviders = serviceProviders.filter(provider => strings.appointmentsServiceProviderRoles.includes(provider.role));
    let appointmentsServiceProvidersTest = [];
    appointmentsServiceProviders.forEach((provider, index) => {
        if (index > 159)
            provider.fullname = "Tara";
        else
            provider.fullname = "Gil";
        appointmentsServiceProvidersTest.push(provider);
    });


    beforeAll(async (done) => {
        wrapper = await shallow(<AppointmentRequest navigation={navigation}/>);
        componentInstance = wrapper.instance();
        await wrapper.update();
        componentInstance.setState({serviceProviders: appointmentsServiceProvidersTest});
        done();
    });

    afterAll(() => {
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    afterEach(() => {
        componentInstance.setState({serviceProviders: appointmentsServiceProvidersTest});
    });

    it('match snapshot', async () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("render with what the user see", async () => {
        expect(wrapper.find(List)).toHaveLength(1);
        expect(wrapper.find('AppointmentRequestForm')).toHaveLength(1);
        expect(wrapper.find(FlatList)).toHaveLength(1);
    });

    it("mounted with the right data", () => {

        expect(componentInstance.state.serviceProviders.length).toEqual(161);
        expect(componentInstance.state.formModal).toEqual(false);
        expect(componentInstance.state.serviceProviderSelected).toEqual({});
        expect(componentInstance.state.noServiceProviderFound).toEqual(false);
    });

    it("search for serviceProviders with name that has ra", async () => {

        const updateSearchSpy = jest.spyOn(componentInstance, 'updateSearch');

        await componentInstance.updateSearch('ra');

        expect(updateSearchSpy).toHaveBeenCalled();
        expect(componentInstance.state.serviceProviders.length).toEqual(1);
    });

    it("search for serviceProviders with role מספרה", async () => {

        const updateSearchSpy = jest.spyOn(componentInstance, 'updateSearch');

        await componentInstance.updateSearch('מספרה');

        expect(updateSearchSpy).toHaveBeenCalled();
        expect(componentInstance.state.serviceProviders.length).toEqual(88);
    });

    it("open AppointmentRequestForm on click serviceProvider", async () => {
        wrapper = shallow(<AppointmentRequest navigation={navigation}/>);
        componentInstance = wrapper.instance();

        const requestAppointmentSpy = jest.spyOn(componentInstance, 'requestAppointment');

        componentInstance.requestAppointment(serviceProviderTest);

        expect(requestAppointmentSpy).toHaveBeenCalled();
        expect(requestAppointmentSpy).toHaveBeenCalledWith(serviceProviderTest);
        expect(componentInstance.state.formModal).toEqual(true);
        expect(componentInstance.state.serviceProviderSelected).toEqual(serviceProviderTest);
    });

    it("close AppointmentRequestForm", async () => {
        wrapper = shallow(<AppointmentRequest navigation={navigation}/>);
        componentInstance = wrapper.instance();

        const closeAppointmentRequestFormSpy = jest.spyOn(componentInstance, 'closeAppointmentRequestForm');

        componentInstance.closeAppointmentRequestForm();

        expect(closeAppointmentRequestFormSpy).toHaveBeenCalled();
        expect(componentInstance.state.formModal).toEqual(false);
        expect(componentInstance.state.serviceProviderSelected).toEqual({});
    });

});
