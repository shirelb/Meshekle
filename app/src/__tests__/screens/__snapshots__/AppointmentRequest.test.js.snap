// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentRequest should match snapshot 1`] = `
<View>
  <List
    containerStyle={
      Object {
        "borderBottomWidth": 0,
        "borderTopWidth": 0,
      }
    }
  >
    <FlatList
      ItemSeparatorComponent={[Function]}
      ListHeaderComponent={[Function]}
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </List>
  <AppointmentRequestForm
    closeAppointmentRequestForm={[Function]}
    modalVisible={false}
    selectedDate="2019-06-03"
    serviceProvider={Object {}}
  />
</View>
`;
